delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/StoragesStatistics.scala

```diff
@@ -0,0 +1,166 @@
+    implicit val keyValueStoreConfig: KeyValueStoreConfig  = config.keyValueStore
+    implicit val persistProgressConfig: SaveProgressConfig = config.persistProgressConfig
+    apply(
+      files
+        .fetch(_, _)
+        .bimap(
+          e => new IllegalStateException(e.reason),
+          _.value.storage.iri
+        ),
+      storages.fetch(_, _).map(_.id),
+      projection,
+      stream
+    )
+  }
+
+  /**
+    * Construct a [[StoragesStatistics]] from a passed ''projection'' and ''stream'' function.
+    * The underlying stream will store its progress and compute the stats for each storage.
+    */
+  private[storages] def apply(
+      fetchFileStorage: (Iri, ProjectRef) => Task[Iri],
+      fetchStorageId: (IdSegment, ProjectRef) => IO[StorageFetchRejection, Iri],
+      projection: Projection[StorageStatsCollection],
+      stream: StreamFromOffset
+  )(implicit
+      uuidF: UUIDF,
+      as: ActorSystem[Nothing],
+      sc: Scheduler,
+      keyValueStoreConfig: KeyValueStoreConfig,
+      persistProgressConfig: SaveProgressConfig
+  ): Task[StoragesStatistics] = {
+
+    val cache =
+      CompositeKeyValueStore[ProjectRef, Iri, StorageStatEntry](
+        id,
+        (_, stats) => stats.lastProcessedEventDateTime.toEpochMilli
+      )
+
+    // Build the storage stat entry from the file event
+    def fileEventToStatEntry(f: FileEvent): Task[((ProjectRef, Iri), StorageStatEntry)] = f match {
+      case c: FileCreated if c.attributes.digest == Digest.NotComputedDigest =>
+        UIO.pure((c.project, c.storage.iri) -> StorageStatEntry(1L, 0L, c.instant))
+      case c: FileCreated                                                    =>
+        UIO.pure((c.project, c.storage.iri) -> StorageStatEntry(1L, c.attributes.bytes, c.instant))
+      case u: FileUpdated                                                    => UIO.pure((u.project, u.storage.iri) -> StorageStatEntry(1L, u.attributes.bytes, u.instant))
+      case fau: FileAttributesUpdated                                        =>
+        fetchFileStorage(fau.id, fau.project).map { storageIri =>
+          (fau.project, storageIri) -> StorageStatEntry(0L, fau.bytes, fau.instant)
+        }
+      case other                                                             =>
+        fetchFileStorage(other.id, other.project).map { storageIri =>
+          (other.project, storageIri) -> StorageStatEntry(0L, 0L, other.instant)
+        }
+    }
+
+    def buildStream: Stream[Task, Unit] =
+      Stream
+        .eval(projection.progress(projectionId))
+        .evalTap { progress =>
+          cache.putAll(progress.value.value)
+        }
+        .flatMap { progress =>
+          val initial = SuccessMessage(progress.offset, progress.timestamp, "", 1, progress.value, Vector.empty)
+          stream(progress.offset)
+            .evalMapAccumulate(initial) { case (acc, envelope) =>
```

Comment:
you seem to discard the second element on the tuple of `evalMapAccumulate`. Wouldn't `evalScan` work?

Reply:
I started with evalScan but it also includes the init element as part of the stream:
https://github.com/typelevel/fs2/blob/main/core/shared/src/main/scala/fs2/Stream.scala#L997

So in the processed count of the projection, there would be +1